### ISSUE TITLE
Why not use generics with the deserialize method definition?

### ISSUE THREAD

**Issue by g00dnatur3** [dhlab-basel/json2typescript#73]
Look here:

```
function deserialize<T>(data: object | [], T: {
	new (): T;
}): T {
	return jsonConvert.deserialize(data, T)
}
```

is there any value to using this approach?

**Comment by andreas-aeschlimann** [comment]
I originally wanted to use only generics instead of the class parameter. The problem is that in TypeScript you cannot write something like new T(). In other languages it is possible though.

This means the parameter with the class is necessary. Using generics is then not of any use because the data is redundant. The only benefit is probably that you may get a proper return type. Or is there something I missed?

**Comment by g00dnatur3** [comment]
I was hopping the generics would solve the issue I am describing below

Lets say I have two classes A & B -- with no getters or setters, just public variables

(I have tried with getters/setters and it soesnt make a difference)

I would expect, with using the generics above (and thus enforcing a return type) would cause the compiler to complain about the code below:

```
const b: B = jsonConvert(data, A)
```
But the compiler does not complain :(

This is the biggest issue I have with Typescript, maybe I am doing something wrong...

But to me these are incompatible types and the assignment should not be allowed.

What do you think?

**Comment by andreas-aeschlimann** [comment]
You dont need generics to force a type. Just use <A> jsonConvert.deserialize(data, A). This tells the compiler that type A is returned.

**Comment by m-architek** [comment]
I've wrapped JsonConvert as follows and it seems to work correctly:

```ts
@Injectable({providedIn: 'root'})
export class JsonMapper extends JsonConvert {

    constructor() {
        super();
        this.operationMode = OperationMode.LOGGING;
        this.ignorePrimitiveChecks = false;
        this.valueCheckingMode = ValueCheckingMode.ALLOW_NULL;
    }

    deserialize<T>(json: any, type: {new (): T}): T {
        return super.deserialize(json, type);
    }
}
```
It makes following code to rise an error:
```ts
const result: Foo = this.jsonMapper.deserialize(obj, Bar); // error
```
I think this kind of type check in original JsonConvert would be great.

**Comment by andreas-aeschlimann** [comment]
I thought about it and yes I agree it is better, I'll add it in the next version!

**Comment by andreas-aeschlimann** [comment]
Hey @g00dnatur3 & @m-architek, just a little question about this issue.

Why do you think is this approach better than just using

const result: Foo = <Foo> this.jsonConvert.deserialize(obj, Foo);

? In the end, the generics are not forced anyway, so you have to remember to type it anyway.

I agree it feels more clean to use generics, but due to the lack of being able to do `new T()` they do not result in any benefit.

**Comment by m-architek** [comment]
In my opinion it is better to have this type check because it is for no cost at all and it can help to proof against some copy-paste mistakes or whatever.

I don't see any reason anyone would ever want to do this:
``` js
const result: Foo = this. jsonConvert.deserialize(obj, Bar);
```

**Comment by g00dnatur3** [comment]
I agree with m-architek, the type-checking should be at compile-time whenever possible...

As I pointed out earlier, in my VS-Code IDE 

`const b: B = jsonConvert(data, A)` 

This should show a red-underline and immediately alert me to the error.. A != B right..

However, this kind of mistake is currently only caught during runtime... by then its too late.

This is just our feedback, we all chose TypeScript because of the type-checking, so it makes sense this is a great feature to have :)

**Comment by andreas-aeschlimann** [comment]
I didn't know that it works without adding the actual generic type to the method. I agree with you.

However...

Some issues arise with this idea:

The basic method `deserialize` currently returns either null, T or T[]. This means that

    const cat: Dog = jsonConvert.deserialize(cat1JsonObject, Dog);

does not work because the compiler will also expect Dog[] or null. Because the deserialize method is used recursively within nested objects, we cannot disallow null or it will break code for people who allow null values for objects or even primitive values.

Your thoughts?
